File: src-tauri/src/commands.rs
use tauri::AppHandle;

use crate::build::{ports::SerialPort, runner};
use crate::codegen::{emitter::GeneratedProject, render};
use crate::identity::registry;
use crate::model::{
    project as proj,
    types::{Board, BoardIdentity, Control, Panel, Project},
    validation::{validate as do_validate, ValidationReport},
};
use crate::pins::allocator::{compute_pin_map, PinMap};

// ── Project file operations ──────────────────────────────────────────────────

#[tauri::command]
pub fn project_new(name: String) -> Result<Project, String> {
    Ok(proj::new_project(name))
}

#[tauri::command]
pub fn project_open(path: String) -> Result<Project, String> {
    proj::load_project(std::path::Path::new(&path)).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn project_save(path: String, project: Project) -> Result<(), String> {
    proj::save_project(std::path::Path::new(&path), &project).map_err(|e| e.to_string())
}

// ── Panel mutations ───────────────────────────────────────────────────────────

#[tauri::command]
pub fn panel_upsert(mut project: Project, panel: Panel) -> Result<Project, String> {
    if let Some(existing) = project.panels.iter_mut().find(|p| p.id == panel.id) {
        *existing = panel;
    } else {
        project.panels.push(panel);
    }
    Ok(project)
}

#[tauri::command]
pub fn panel_delete(mut project: Project, panel_id: String) -> Result<Project, String> {
    project.panels.retain(|p| p.id != panel_id);
    project.controls.retain(|c| c.panel_id() != panel_id);
    Ok(project)
}

// ── Board mutations ────────────────────────────────────────────────────────────

#[tauri::command]
pub fn board_upsert(mut project: Project, board: Board) -> Result<Project, String> {
    if let Some(existing) = project.boards.iter_mut().find(|b| b.id == board.id) {
        *existing = board;
    } else {
        project.boards.push(board);
    }
    Ok(project)
}

#[tauri::command]
pub fn board_delete(mut project: Project, board_id: String) -> Result<Project, String> {
    project.boards.retain(|b| b.id != board_id);
    project.controls.retain(|c| c.board_id() != board_id);
    Ok(project)
}

// ── Control mutations ──────────────────────────────────────────────────────────

#[tauri::command]
pub fn control_upsert(mut project: Project, control: Control) -> Result<Project, String> {
    let id = control.id().to_string();
    if let Some(existing) = project.controls.iter_mut().find(|c| c.id() == id) {
        *existing = control;
    } else {
        project.controls.push(control);
    }
    Ok(project)
}

#[tauri::command]
pub fn control_delete(mut project: Project, control_id: String) -> Result<Project, String> {
    project.controls.retain(|c| c.id() != control_id);
    Ok(project)
}

// ── Validation & pin allocation ────────────────────────────────────────────────

#[tauri::command]
pub fn validate(project: Project) -> Result<ValidationReport, String> {
    Ok(do_validate(&project))
}

#[tauri::command]
pub fn board_pinmap(project: Project, board_id: String) -> Result<PinMap, String> {
    Ok(compute_pin_map(&project, &board_id))
}

#[tauri::command]
pub fn allocate_identity(
    mut project: Project,
    board_id: String,
) -> Result<(Project, BoardIdentity), String> {
    let identity = registry::allocate_identity(&mut project, &board_id)
        .map_err(|e| e.to_string())?;
    Ok((project, identity))
}

// ── Codegen ────────────────────────────────────────────────────────────────────

#[tauri::command]
pub fn generate_board(project: Project, board_id: String) -> Result<GeneratedProject, String> {
    let generated_board =
        render::render_board(&project, &board_id).map_err(|e| e.to_string())?;
    Ok(crate::codegen::emitter::to_generated_project(
        &board_id,
        generated_board,
    ))
}

// ── Serial ports ───────────────────────────────────────────────────────────────

#[tauri::command]
pub fn list_serial_ports() -> Result<Vec<SerialPort>, String> {
    Ok(crate::build::ports::list_serial_ports())
}

// ── Build & upload ─────────────────────────────────────────────────────────────

#[tauri::command]
pub async fn build_board(
    app: AppHandle,
    project: Project,
    board_id: String,
    port: Option<String>,
) -> Result<(), String> {
    runner::build_board(app, project, board_id, port)
        .await
        .map_err(|e| e.to_string())
}

File: electron/engine/commands.ts
// Command surface — ported from server/src/commands.rs. Pure functions over a
// Project; mutations return an updated copy (the renderer holds the source of truth).

import type {
  Board,
  BoardIdentity,
  ButtonControl,
  Control,
  EncoderControl,
  GeneratedProject,
  Panel,
  PinMap,
  Project,
  SelectorControl,
  SwitchControl,
  AnalogControl,
  ValidationReport,
} from "./types";
import { newProject, parseProject, serializeProject } from "./model";
import { validate } from "./validation";
import { computePinMap } from "./pins";
import { allocateIdentity as allocate } from "./identity";
import { renderBoard } from "./render";
import { toGeneratedProject } from "./emitter";

const clone = <T>(v: T): T => structuredClone(v);

// ── Project file operations ──────────────────────────────────────────────────

export const projectNew = (name: string): Project => newProject(name);
export const projectOpen = (content: string): Project => parseProject(content);
export const projectSerialize = (project: Project): string => serializeProject(project);

// ── Panel mutations ───────────────────────────────────────────────────────────

export function panelUpsert(project: Project, panel: Panel): Project {
  const next = clone(project);
  const i = next.panels.findIndex((p) => p.id === panel.id);
  if (i >= 0) next.panels[i] = panel;
  else next.panels.push(panel);
  return next;
}

export function panelDelete(project: Project, panelId: string): Project {
  const next = clone(project);
  next.panels = next.panels.filter((p) => p.id !== panelId);
  next.controls = next.controls.filter((c) => c.panelId !== panelId);
  return next;
}

// ── Board mutations ────────────────────────────────────────────────────────────

export function boardUpsert(project: Project, board: Board): Project {
  const next = clone(project);
  const i = next.boards.findIndex((b) => b.id === board.id);
  if (i >= 0) next.boards[i] = board;
  else next.boards.push(board);
  return next;
}

export function boardDelete(project: Project, boardId: string): Project {
  const next = clone(project);
  next.boards = next.boards.filter((b) => b.id !== boardId);
  next.controls = next.controls.map((c) => (c.boardId === boardId ? unassign(c) : c));
  return next;
}

/** Clear a control's board + kind-specific pin config, leaving it "unassigned". */
function unassign(control: Control): Control {
  const base = { ...control, boardId: undefined };
  switch (control.kind) {
    case "button":
      return { ...base, pin: undefined } as ButtonControl;
    case "switch":
      return { ...base, pin: undefined } as SwitchControl;
    case "selector":
      return { ...base, positions: [] } as SelectorControl;
    case "encoder":
      return { ...base, encoder: undefined } as EncoderControl;
    case "analog":
      return { ...base, analog: undefined } as AnalogControl;
  }
}

// ── Control mutations ──────────────────────────────────────────────────────────

export function controlUpsert(project: Project, control: Control): Project {
  const next = clone(project);
  const i = next.controls.findIndex((c) => c.id === control.id);
  if (i >= 0) next.controls[i] = control;
  else next.controls.push(control);
  return next;
}

export function controlDelete(project: Project, controlId: string): Project {
  const next = clone(project);
  next.controls = next.controls.filter((c) => c.id !== controlId);
  return next;
}

// ── Validation & pin allocation ────────────────────────────────────────────────

export const validateProject = (project: Project): ValidationReport => validate(project);

export const boardPinmap = (project: Project, boardId: string): PinMap =>
  computePinMap(project, boardId);

export function allocateIdentity(
  project: Project,
  boardId: string,
): [Project, BoardIdentity] {
  const next = clone(project);
  const identity = allocate(next, boardId);
  return [next, identity];
}

// ── Codegen ────────────────────────────────────────────────────────────────────

export function generateBoard(project: Project, boardId: string): GeneratedProject {
  const generated = renderBoard(project, boardId);
  return toGeneratedProject(boardId, generated);
}
